File: 31-next-permutation/31-next-permutation.cs
public class Solution {
  public void NextPermutation(int[] nums) {
    int k = nums.Length - 2;
    for (; k >= 0; k--) {
      if (nums[k+1] > nums[k]) break;
    }
    if (k >= 0) {
      int l = nums.Length - 1;
      int v = nums[k];
      for (; l > k + 1; l--) {
        if (nums[l] > v) break;
      }
      nums[k] = nums[l];
      nums[l] = v;
    }
    for (int i = k+1, j = nums.Length-1; i < j; i++, j--) {
      (nums[i], nums[j]) = (nums[j], nums[i]);
    }
  }
}

File: 31-next-permutation/31-next-permutation.rs
impl Solution {
  pub fn next_permutation(nums: &mut Vec<i32>) {
    let n:i32 = nums.len() as i32;
    let mut k:i32 = n - 2;
    while k >= 0 && nums[k as usize] >= nums[(k+1) as usize] {
      k -= 1;
    }
    if k < 0 {
      nums.reverse();
      return
    }
    let mut l = n - 1;
    let v = nums[k as usize];
    while l > k && nums[l as usize] <= v {
      l -= 1;
    }
    nums[k as usize] = nums[l as usize];
    nums[l as usize] = v;
    let mut i = k+1;
    let mut j = n-1;
    while i < j {
      nums.swap(i as usize, j as usize);
      i += 1;
      j -= 1;
    }
  }
}